README.md
# iocage-provision

Creates an iocage based FreeBSD jail.

|                  |                                                         |
| ---------------: | ------------------------------------------------------- |
|               CI | [![CI Status][badge-ci-overall]][ci]                    |
|   Latest Version | [![Latest version][badge-version]][crate]               |
|    Documentation | [![Documentation][badge-docs]][docs]                    |
|  Crate Downloads | [![Crate downloads][badge-crate-dl]][crate]             |
| GitHub Downloads | [![GitHub downloads][badge-github-dl]][github-releases] |
|          License | [![Crate license][badge-license]][github]               |

This program uses [iocage] to create a VNET networked [ZFS]-backed [FreeBSD]
jail. Suitable defaults are computed for the default gateway and base release to
reduce the number of arguments in the common case. An optional `--ssh` flag will
install and start an SSH service when the jail boots for remote management.
Finally, an optional `--user` option will create a user in the new jail by
copying values from the outside/host system.

[freebsd]: https://www.freebsd.org/
[iocage]: https://iocage.io/
[zfs]: https://zfsonfreebsd.github.io/ZoF/

**Table of Contents**

<!-- toc -->

- [Installation](#installation)
  - [Cargo Install](#cargo-install)
  - [From Source](#from-source)
- [Usage](#usage)
  - [Example 1 Provisioning a New Jail With a Name and Address](#example-1-provisioning-a-new-jail-with-a-name-and-address)
  - [Example 2 Provisioning a New Jail With a User and SSH Service](#example-2-provisioning-a-new-jail-with-a-user-and-ssh-service)
  - [Example 3 Using a Custom Default Gateway and Base Release](#example-3-using-a-custom-default-gateway-and-base-release)
- [CI Status](#ci-status)
  - [Build (main branch)](#build-main-branch)
  - [Test (main branch)](#test-main-branch)
  - [Check (main branch)](#check-main-branch)
- [Code of Conduct](#code-of-conduct)
- [Issues](#issues)
- [Contributing](#contributing)
- [Release History](#release-history)
- [Authors](#authors)
- [License](#license)

<!-- tocstop -->

## Installation

This program is intended to run on FreeBSD systems only, so while it may compile
on other systems, it likely won't do what you expect ;)

### Cargo Install

If [Rust](https://rustup.rs/) is installed, then installing with Cargo is
straight forward:

```console
$ cargo install iocage-provision
```

### From Source

To install from source, you can clone the Git repository, build with Cargo and
copy the binary into a destination directory. This will build the project from
the latest commit on the main branch, which may not correspond to the latest
stable release:

```console
$ git clone https://github.com/fnichol/iocage-provision.git
$ cd iocage-provision
$ cargo build --release
$ cp ./target/release/iocage-provision /dest/path/
```

## Usage

### Example 1 Provisioning a New Jail With a Name and Address

The following command will create a new jail called `ferris` with an IP
address/subnet mask of `192.168.0.100/24`.

```console
$ iocage-provision ferris 192.168.0.100/24
```

### Example 2 Provisioning a New Jail With a User and SSH Service

The following command will create a new jail with a running SSH service, and a
user called `jdoe` which is copied from the host system (note that the user must
exist on the host system).

```console
$ iocage-provision --user jdoe --ssh homebase 10.0.0.25/24
```

### Example 3 Using a Custom Default Gateway and Base Release

The following command will create a new jail by overriding the default gateway
and default base release values.

```console
$ iocage-provision --gateway 10.1.0.254 --release 11.1-RELEASE \
  bespoke 10.1.0.1/24
```

## CI Status

### Build (main branch)

| Operating System | Stable Rust                                                             |
| ---------------: | ----------------------------------------------------------------------- |
|          FreeBSD | [![FreeBSD Stable Build Status][badge-stable_freebsd-build]][ci-main] |

### Test (main branch)

| Operating System | Stable Rust                                                           |
| ---------------: | --------------------------------------------------------------------- |
|          FreeBSD | [![FreeBSD Stable Test Status][badge-stable_freebsd-test]][ci-main] |

### Check (main branch)

|        | Status                                            |
| ------ | ------------------------------------------------- |
| Lint   | [![Lint Status][badge-check-lint]][ci-main]     |
| Format | [![Format Status][badge-check-format]][ci-main] |

## Code of Conduct

This project adheres to the Contributor Covenant [code of
conduct][code-of-conduct]. By participating, you are expected to uphold this
code. Please report unacceptable behavior to fnichol@nichol.ca.

## Issues

If you have any problems with or questions about this project, please contact us
through a [GitHub issue][issues].

## Contributing

You are invited to contribute to new features, fixes, or updates, large or
small; we are always thrilled to receive pull requests, and do our best to
process them as fast as we can.

Before you start to code, we recommend discussing your plans through a [GitHub
issue][issues], especially for more ambitious contributions. This gives other
contributors a chance to point you in the right direction, give you feedback on
your design, and help you find out if someone else is working on the same thing.

## Release History

See the [changelog] for a full release history.

## Authors

Created and maintained by [Fletcher Nichol][fnichol] (<fnichol@nichol.ca>).

## License

Licensed under the Mozilla Public License Version 2.0 ([LICENSE.txt][license]).

Unless you explicitly state otherwise, any contribution intentionally submitted
for inclusion in the work by you, as defined in the MPL-2.0 license, shall be
licensed as above, without any additional terms or conditions.

[badge-check-format]:
  https://img.shields.io/cirrus/github/fnichol/iocage-provision.svg?style=flat-square&task=check&script=format
[badge-check-lint]:
  https://img.shields.io/cirrus/github/fnichol/iocage-provision.svg?style=flat-square&task=check&script=lint
[badge-ci-overall]:
  https://img.shields.io/cirrus/github/fnichol/iocage-provision.svg?style=flat-square
[badge-crate-dl]:
  https://img.shields.io/crates/d/iocage-provision.svg?style=flat-square
[badge-docs]: https://docs.rs/iocage-provision/badge.svg?style=flat-square
[badge-github-dl]:
  https://img.shields.io/github/downloads/fnichol/iocage-provision/total.svg?style=flat-square
[badge-license]:
  https://img.shields.io/crates/l/iocage-provision.svg?style=flat-square
[badge-stable_freebsd-build]:
  https://img.shields.io/cirrus/github/fnichol/iocage-provision.svg?style=flat-square&task=test_stable_freebsd&script=build
[badge-stable_freebsd-test]:
  https://img.shields.io/cirrus/github/fnichol/iocage-provision.svg?style=flat-square&task=test_stable_freebsd&script=test
[badge-version]:
  https://img.shields.io/crates/v/iocage-provision.svg?style=flat-square
[changelog]:
  https://github.com/fnichol/iocage-provision/blob/main/CHANGELOG.md
[ci]: https://cirrus-ci.com/github/fnichol/iocage-provision
[ci-main]: https://cirrus-ci.com/github/fnichol/iocage-provision/main
[code-of-conduct]:
  https://github.com/fnichol/iocage-provision/blob/main/CODE_OF_CONDUCT.md
[crate]: https://crates.io/crates/iocage-provision
[docs]: https://docs.rs/iocage-provision
[fnichol]: https://github.com/fnichol
[github-releases]: https://github.com/fnichol/iocage-provision/releases
[github]: https://github.com/fnichol/iocage-provision
[issues]: https://github.com/fnichol/iocage-provision/issues
[license]: https://github.com/fnichol/iocage-provision/blob/main/LICENSE.txt


CHANGELOG.md
# Changelog

<!-- next-header -->

## [Unreleased] - ReleaseDate

## [0.2.0] - 2021-07-04

### Added

- add optional support for "thick jails"

### Changed

- refactor indented line output code

## [0.1.1] - 2021-07-03

## [0.1.0] - 2021-07-02

- The initial release.

<!-- next-url -->

[unreleased]: https://github.com/fnichol/iocage-provision/compare/v0.2.0...HEAD

[0.2.0]: https://github.com/fnichol/iocage-provision/compare/v0.1.1...v0.2.0
[0.1.1]: https://github.com/fnichol/iocage-provision/compare/v0.1.0...v0.1.1
[0.1.0]: https://github.com/fnichol/fnichol-cime/compare/d9f9b0d...v0.1.0


tests/version-numbers.rs
// Copyright 2020 Fletcher Nichol and/or applicable contributors.
//
// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license (see <LICENSE-MIT or
// http://opensource.org/licenses/MIT>, at your option. This file may not be copied, modified, or
// distributed except according to those terms.

#[test]
fn test_readme_deps() {
    // If the current version is a `*-dev` string, then ignore the check in README. We'd like to
    // keep the *last* released version string in the README as instructions in the main source
    // code branch.
    if !env!("CARGO_PKG_VERSION").ends_with("-dev") {
        version_sync::assert_markdown_deps_updated!("README.md");
    }
}

#[test]
fn test_html_root_url() {
    version_sync::assert_html_root_url_updated!("src/lib.rs");
}


build.rs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

use inner::Info;
use std::env;
use std::fs::File;
use std::io::{self, BufWriter};
use std::path::Path;

fn main() {
    let mut info = Info::new();

    generate_version_short(&mut info);
    generate_version_long(&mut info);

    println!("cargo:rerun-if-env-changed=NIGHTLY_BUILD");
    if let Ok(date) = std::env::var("NIGHTLY_BUILD") {
        println!(
            "cargo:rustc-env=CARGO_PKG_VERSION={}-nightly.{}",
            std::env::var("CARGO_PKG_VERSION")
                .unwrap()
                .split('-')
                .next()
                .unwrap(),
            date,
        );
    }
}

fn generate_version_short(info: &mut Info) {
    let dst =
        File::create(Path::new(&env::var("OUT_DIR").unwrap()).join("version_short.txt")).unwrap();
    let mut dst = BufWriter::new(dst);
    io::copy(&mut io::Cursor::new(version_short(info)), &mut dst).unwrap();
}

fn generate_version_long(info: &mut Info) {
    let dst =
        File::create(Path::new(&env::var("OUT_DIR").unwrap()).join("version_long.txt")).unwrap();
    let mut dst = BufWriter::new(dst);
    io::copy(&mut io::Cursor::new(version_long(info)), &mut dst).unwrap();
}

fn version_short(info: &mut Info) -> String {
    let mut version = env!("CARGO_PKG_VERSION").to_string();
    if let Some(hash) = info.commit_hash_short() {
        let mut extra = String::new();
        extra.push_str(" (");
        extra.push_str(&hash);
        extra.push(' ');
        if let Some(date) = info.commit_date() {
            extra.push_str(&date);
            extra.push(')');
            version.push_str(&extra)
        }
    }
    version
}

fn version_long(info: &mut Info) -> String {
    let mut version = version_short(info);
    version.push_str("\nbinary: ");
    version.push_str(env!("CARGO_PKG_NAME"));
    version.push_str("\nrelease: ");
    version.push_str(env!("CARGO_PKG_VERSION"));
    if let Some(hash) = info.commit_hash_long() {
        version.push_str("\ncommit-hash: ");
        version.push_str(&hash);
    }
    if let Some(date) = info.commit_date() {
        version.push_str("\ncommit-date: ");
        version.push_str(&date);
    }
    version
}

mod inner {
    use std::collections::HashMap;
    use std::env;
    use std::process::Command;

    pub struct Info(HashMap<&'static str, Option<String>>);

    impl Info {
        pub fn new() -> Self {
            Info(HashMap::new())
        }

        pub fn commit_hash_short(&mut self) -> Option<&str> {
            self.0
                .entry("commit_hash_short")
                .or_insert_with(commit_hash_short)
                .as_deref()
        }

        pub fn commit_hash_long(&mut self) -> Option<&str> {
            self.0
                .entry("commit_hash_long")
                .or_insert_with(commit_hash_long)
                .as_deref()
        }

        pub fn commit_date(&mut self) -> Option<&str> {
            self.0
                .entry("commit_date")
                .or_insert_with(commit_date)
                .as_deref()
        }
    }

    pub fn commit_hash_short() -> Option<String> {
        let hash = command_stdout(Command::new(git()).args(&["show", "-s", "--format=%h"]));

        match is_dirty() {
            Some(id) if id => hash.map(|hash| format!("{}-dirty", hash)),
            _ => hash,
        }
    }

    pub fn commit_hash_long() -> Option<String> {
        let hash = command_stdout(Command::new(git()).args(&["show", "-s", "--format=%H"]));

        match is_dirty() {
            Some(id) if id => hash.map(|hash| format!("{}-dirty", hash)),
            _ => hash,
        }
    }

    pub fn commit_date() -> Option<String> {
        command_stdout(Command::new(git()).args(&["show", "-s", "--format=%ad", "--date=short"]))
    }

    pub fn is_dirty() -> Option<bool> {
        Command::new(git())
            .args(&["diff-index", "--quiet", "HEAD"])
            .status()
            .ok()
            .map(|status| !status.success())
    }

    fn git() -> String {
        env::var("GIT_CMD").unwrap_or_else(|_| "git".to_string())
    }

    fn command_stdout(cmd: &mut Command) -> Option<String> {
        cmd.output()
            .ok()
            .map(|output| String::from_utf8_lossy(&output.stdout).trim().to_owned())
    }
}


src/main.rs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

use iocage_provision::Result;
use log::{debug, error};
use std::process;

mod cli;

fn main() {
    cli::util::setup_panic_hooks();

    if let Err(err) = try_main() {
        for line in cli::util::pretty_error(&err).lines() {
            error!("{}", line);
        }
        process::exit(1);
    }
}

fn try_main() -> Result<()> {
    let args = cli::from_args();
    cli::util::init_logger(args.verbose);
    debug!("parsed cli arguments; args={:?}", args);

    iocage_provision::ensure_root()?;
    iocage_provision::provision_jail(
        &args.name,
        &args.ip,
        &args.gateway,
        &args.release,
        args.user.as_deref(),
        args.ssh,
    )?;

    Ok(())
}


.ci/cargo-cache-install.sh
hellcheck shell=sh disable=SC2039

print_usage() {
  local program="$1"

  echo "$program

    Installs a Cargo plugin into a dedicated directory for later caching

    USAGE:
        $program [FLAGS] [--] <PLUGIN>

    FLAGS:
        -h, --help      Prints help information

    ARGS:
        <PLUGIN>  Name of the Cargo plugin
    " | sed 's/^ \{1,4\}//g'
}

main() {
  set -eu
  if [ -n "${DEBUG:-}" ]; then set -v; fi
  if [ -n "${TRACE:-}" ]; then set -xv; fi

  local program
  program="$(basename "$0")"

  OPTIND=1
  while getopts "h-:" arg; do
    case "$arg" in
      h)
        print_usage "$program"
        return 0
        ;;
      -)
        case "$OPTARG" in
          help)
            print_usage "$program"
            return 0
            ;;
          '')
            # "--" terminates argument processing
            break
            ;;
          *)
            print_usage "$program" >&2
            fail "invalid argument --$OPTARG"
            ;;
        esac
        ;;
      \?)
        print_usage "$program" >&2
        fail "invalid argument; arg=-$OPTARG"
        ;;
    esac
  done
  shift "$((OPTIND - 1))"

  if [ -z "${1:-}" ]; then
    print_usage "$program" >&2
    fail "missing <PLUGIN> argument; arg=-$OPTARG"
  fi
  local plugin="$1"
  shift

  local root="$CARGO_HOME/opt/$plugin"

  install_plugin "$plugin" "$root"
}

install_plugin() {
  local plugin="$1"
  local root="$2"

  mkdir -p "$root"
  rustup install stable
  cargo +stable install --root "$root" --force --verbose "$plugin"

  # Create symbolic links for all execuatbles into $CARGO_HOME/bin
  ln -snf "$root/bin"/* "$CARGO_HOME/bin/"
}

fail() {
  echo "" >&2
  echo "xxx $1" >&2
  echo "" >&2
  return 1
}

main "$@"


src/cli.rs
// This Source Code Form is subject to the terms of the Mozilla Public
// License, v. 2.0. If a copy of the MPL was not distributed with this
// file, You can obtain one at https://mozilla.org/MPL/2.0/.

use ipnet::IpNet;
use nix::sys::utsname;
use std::error;
use std::fmt;
use std::io;
use std::net::{self, IpAddr};
use std::process::Command;
use std::str;
use structopt::{clap, StructOpt};

lazy_static::lazy_static! {
    /// The computed default value for the gateway option.
    static ref DEFAULT_GATEWAY: String = default_gateway();

    /// The computed default value for the release option.
    static ref DEFAULT_RELEASE: String = default_release();
}

/// The "author" string for help messages.
const AUTHOR: &str = concat!(env!("CARGO_PKG_AUTHORS"), "\n\n");

/// The "about" string for help messages
const ABOUT: &str = concat!(
    "\
Creates an iocage based FreeBSD jail.

Project home page: ",
    env!("CARGO_PKG_HOMEPAGE"),
    "

Use -h for short descriptions and --help for more details.",
);

/// The "long_about" string for help messages.
const LONG_ABOUT: &str = concat!(
    "\
Creates an iocage based FreeBSD jail.

This program uses iocage to create a VNET networked ZFS-backed FreeBSD jail. Suitable defaults are
computed for the default gateway and base release to reduce the number of arguments in the common
case. An optional --ssh flag will install and start an SSH service when the jail boots for remote
management. Finally, an optional --user option will create a user in the new jail by copying values
from the outside/host system.

Project home page: ",
    env!("CARGO_PKG_HOMEPAGE"),
    "

Use -h for short descriptions and --help for more details.",
);

/// An examples section at the end of the help message.
const AFTER_HELP: &str = "\
EXAMPLES:
    Example 1 Provisioning a New Jail With a Name and Address

      The following command will create a new jail called ferris with an IP
      address/subnet mask of 192.168.0.100/24.

        # iocage-provision ferris 192.168.0.100/24

    Example 2 Provisioning a New Jail With a User and SSH Service

      The following command will create a new jail with a running SSH service,
      and a user called jdoe which is copied from the host system (note that
      the user must exist on the host system).

        # iocage-provision --user jdoe --ssh homebase 10.0.0.25/24

    Example 3 Using a Custom Default Gateway and Base Release

      The following command will create a new jail by overriding the default
      gateway and default base release values.

        # iocage-provision --gateway 10.1.0.254 --release 11.1-RELEASE \\
          bespoke 10.1.0.1/24

";

/// Parse, validate, and return the CLI arguments as a typed struct.
pub(crate) fn from_args() -> Args {
    Args::from_clap(
        &Args::clap()
            // TODO: StructOpt generates a Clap app with an unconditional
            // `.version(env!("CARGO_PKG_VERSION"))` at the end of the builder chain which
            // overrides any values inserted in the proc macro. Until this behavior can be
            // fixed, this is a temporary workaround which wraps the underlying `App` type and
            // chains on a call to `version`.
            .version(BuildInfo::version_short())
            .get_matches(),
    )
}

/// The parsed CLI arguments.
#[derive(Debug, StructOpt)]
#[structopt(
    global_settings(&[clap::AppSettings::UnifiedHelpMessage]),
    max_term_width = 100,
    author = AUTHOR,
    about = ABOUT,
    long_about = LONG_ABOUT,
    version = BuildInfo::version_short(),
    long_version = BuildInfo::version_long(),
    after_help = AFTER_HELP,
)]
pub(crate) struct Args {
    /// IP address of the default gateway route for a VNET.
    ///
    /// This address is used when setting up the VNET networking of the jail. If not provided the
    /// default value will be the address corresponding to the default route on the underlying host
    /// as determined by using the `netstat` program.
    #[structopt(
        short = "g",
        long,
        default_value = &DEFAULT_GATEWAY,
        rename_all = "screaming-snake",
    )]
    pub(crate) gateway: IpAddr,

    /// IP address & subnet mask for the jail instance. [example: 10.200.0.50/24]
    ///
    /// The IP address and the subnet mask are both required for the value to be considered valid.
    #[structopt(index = 2, rename_all = "screaming-snake")]
    pub(crate) ip: IpNet,

    /// Name for the jail instance [example: myjail]
    #[structopt(index = 1, rename_all = "screaming-snake")]
    pub(crate) name: String,

    /// FreeBSD release to use for the jail instance.
    ///
    /// If not provided, the default value will be the same release version that is running on the
    /// underlying host system. For example if `uname -r` returns `11.2-STABLE`, then the default
    /// value would be `11.2-RELEASE`.
    #[structopt(
        short = "R",
        long,
        default_value = &DEFAULT_RELEASE,
        rename_all = "screaming-snake",
    )]
    pub(crate) release: String,

    /// Installs and sets up an SSH service.
    ///
    /// If this flag is set, then SSH software is installed, enabled on boot and is started on
    /// first boot. Useful for jails that required remote administration, remote file copying, etc.
    #[structopt(short = "s", long)]
    pub(crate) ssh: bool,

    /// User to create in jail instance (based on host system's information).
    ///
    /// When this option is used, a user account will be created in the new jail with settings
    /// copied from the underlying system's `passwd` database. In other words, the username
    /// provided must exist on the host system, otherwise the command will result in an error and
    /// the jail will not be created.
    #[structopt(short = "u", long, rename_all = "screaming-snake")]
    pub(crate) user: Option<String>,

    /// Sets the verbosity mode.
    ///
    /// Multiple -v options increase verbosity. The maximum is 3.
    #[structopt(short = "v", long = "verbose", parse(from_occurrences))]
    pub(crate) verbose: usize,
}

/// A default gateway value.
fn default_gateway() -> String {
    netstat_gateway_addr()
        .unwrap_or_else(|err| {
            clap::Error::with_description(
                &format!("could not determine default gateway, {}", err.to_string()),
                clap::ErrorKind::Io,
            )
            .exit()
        })
        .to_string()
}

/// A default release value.
fn default_release() -> String {
    utsname::uname().release().replace("-STABLE", "-RELEASE")
}

/// Determines and returns a default gateway IP address by querying the `netstat` command.
///
/// # Errors
///
/// Returns an `Err` if:
///
/// * The `netstat` command cannot be found
/// * The output of the command cannot be parsed as UTF-8
/// * No line of output starting with `"default"` can be found
/// * The default line cannot be successfully split
/// * The IP address string cannot be parsed as an IP address
fn netstat_gateway_addr() -> Result<IpAddr, GatewayError> {
    str::from_utf8(
        Command::new("netstat")
            .args(&["-r", "-n", "-f", "inet"])
            .output()
            .map_err(GatewayError::Cmd)?
            .stdout
            .as_ref(),
    )
    .map_err(GatewayError::Utf8)?
    .lines()
    .find(|line| line.starts_with("default"))
    .map(|line| line.split_ascii_whitespace())
    .ok_or(GatewayError::NetstatParse("netstat default line not found"))?
    .nth(1)
    .ok_or(GatewayError::NetstatParse(
        "netstat second column not found on default line",
    ))?
    .parse()
    .map_err(GatewayError::IpAddr)
}

/// Error when determining a default gateway IP address.
#[derive(Debug)]
enum GatewayError {
    /// A command cannot be found or run successfully.
    Cmd(io::Error),
    /// An IP address failed to be parsed.
    IpAddr(net::AddrParseError),
    /// The output of the `netstat` command failed to be parsed.
    NetstatParse(&'static str),
    /// A string failed to be parsed as UTF-8.
    Utf8(str::Utf8Error),
}

impl fmt::Display for GatewayError {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        use GatewayError::*;

        match self {
            Cmd(ref err) => write!(f, "failed to run netstat command ({})", err),
            IpAddr(ref err) => err.fmt(f),
            NetstatParse(ref msg) => f.write_str(msg),
            Utf8(ref err) => err.fmt(f),
        }
    }
}

impl error::Error for GatewayError {
    fn source(&self) -> Option<&(dyn error::Error + 'static)> {
        use GatewayError::*;

        match self {
            Cmd(ref err) => err.source(),
            IpAddr(ref err) => err.source(),
            NetstatParse(_) => None,
            Utf8(ref err) => err.source(),
        }
    }
}

/// Build time metadata.
struct BuildInfo;

impl BuildInfo {
    /// Returns a short version string.
    fn version_short() -> &'static str {
        include_str!(concat!(env!("OUT_DIR"), "/version_short.txt"))
    }

    /// Returns a long version string.
    fn version_long() -> &'static str {
        include_str!(concat!(env!("OUT_DIR"), "/version_long.txt"))
    }
}

pub(crate) mod util {
    use chrono::{SecondsFormat, Utc};
    use std::env;
    use std::panic;

    /// The logger.
    const LOGGER: &Logger = &Logger;

    /// A custom and minimal `Log` implementation.
    ///
    /// This logger writes either to the standard output stream or standard error stream, depending
    /// on the log level.
    ///
    /// Thanks to the logger implementations from ripgrep and the simplelog crate which served as
    /// an inspiration.
    struct Logger;

    impl log::Log for Logger {
        fn enabled(&self, _: &log::Metadata) -> bool {
            true
        }

        fn log(&self, record: &log::Record) {
            if log::max_level() == log::LevelFilter::Info {
                match record.level() {
                    log::Level::Info => println!("  - {}", record.args()),
                    log::Level::Warn => eprintln!("!!! {}", record.args()),
                    log::Level::Error => eprintln!("xxx {}", record.args()),
                    _ => unreachable!("illegal log level"),
                }
            } else {
                let file = record.file().unwrap_or("<unknown>");
                let location = match record.line() {
                    Some(line) => format!("{}:{}", file, line),
                    None => format!("{}:<unknown>", file),
                };

                match record.level() {
                    log::Level::Info => {
                        println!(
                            "{} {:<5} [{}] {}",
                            Utc::now().to_rfc3339_opts(SecondsFormat::Secs, true),
                            record.level(),
                            location,
                            record.args()
                        );
                    }
                    _ => {
                        eprintln!(
                            "{} {:<5} [{}] {}",
                            Utc::now().to_rfc3339_opts(SecondsFormat::Secs, true),
                            record.level(),
                            location,
                            record.args()
                        );
                    }
                }
            }
        }

        fn flush(&self) {
            // `eprintln!` and `println!` flush on every call
        }
    }

    /// Sets up and initializes the logger.
    pub(crate) fn init_logger(verbosity: usize) {
        log::set_logger(LOGGER).expect("error setting logger");

        match verbosity {
            0 => log::set_max_level(log::LevelFilter::Info),
            1 => log::set_max_level(log::LevelFilter::Debug),
            v if v >= 2 => log::set_max_level(log::LevelFilter::Trace),
            _ => {}
        }
        log::debug!("verbosity={}", verbosity);
    }

    /// Wires up a human-first experience if the program panics unexpectedly and also supports the
    /// normal `RUST_BACKTRACE` environment variable.
    ///
    /// A big thanks to https://github.com/rustwasm/wasm-pack for demonstrating such a delightful
    /// pattern. All credit here goes to the wasm-pack crew, thanks!
    pub(crate) fn setup_panic_hooks() {
        let meta = human_panic::Metadata {
            version: env!("CARGO_PKG_VERSION").into(),
            name: env!("CARGO_PKG_NAME").into(),
            authors: env!("CARGO_PKG_AUTHORS").into(),
            homepage: env!("CARGO_PKG_HOMEPAGE").into(),
        };

        let default_hook = panic::take_hook();

        if env::var("RUST_BACKTRACE").is_err() {
            panic::set_hook(Box::new(move |info: &panic::PanicInfo| {
                // First call the default hook that prints to standard error
                default_hook(info);

                // Then call human panic
                let file_path = human_panic::handle_dump(&meta, info);
                human_panic::print_msg(file_path, &meta)
                    .expect("human-panic: printing error message to console failed");
            }));
        }
    }

    /// Return a prettily formatted error, including its entire causal chain.
    ///
    /// Thanks again to the imdb-rename crate and wasm-pack which form the basis of this
    /// implementation.
    pub(crate) fn pretty_error(err: &failure::Error) -> String {
        let mut pretty = "Error: ".to_string();
        pretty.push_str(&err.to_string());
        pretty.push('\n');
        for cause in err.iter_causes() {
            pretty.push_str("Caused by: ");
            pretty.push_str(&cause.to_string());
            pretty.push('\n');
        }
        pretty
    }
}
